File: diploma-server/commons/src/test/java/org/arhor/diploma/commons/IntArray2DTest.java
package org.arhor.diploma.commons;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.Random;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

class IntArray2DTest {

    @Test
    void shouldCreateEmptyArrayWithDefinedSize() {
        // given
        final int rows = 3;
        final int cols = 3;

        // when
        final var array = IntArray2D.create(rows, cols);

        // then
        assertThat(array).isNotNull();
        assertThat(array.rows).isEqualTo(rows);
        assertThat(array.cols).isEqualTo(cols);

        for (int i = 0; i < array.rows; i++) {
            for (int j = 0; j < array.cols; j++) {
                assertThat(array.get(i, j)).isZero();
            }
        }
    }

    @Test
    void shouldCreateFilledArrayWithDefinedSize() {
        // given
        final int rows = 3;
        final int cols = 3;

        // when
        final var array = IntArray2D.create(rows, cols, (col, row) -> 1);

        // then
        assertThat(array).isNotNull();
        assertThat(array.rows).isEqualTo(rows);
        assertThat(array.cols).isEqualTo(cols);

        for (int i = 0; i < array.rows; i++) {
            for (int j = 0; j < array.cols; j++) {
                assertThat(array.get(i, j)).isOne();
            }
        }
    }

    @ParameterizedTest
    @MethodSource("provideArrayAndRandomCell")
    void shouldSetExactlyOneCell(int rows, int cols, int testRow, int testCol) {
        // when
        final var array = IntArray2D.create(rows, cols);
        array.set(testRow, testCol, 1);

        // then
        assertThat(array)
                .isNotNull()
                .hasFieldOrPropertyWithValue("rows", rows)
                .hasFieldOrPropertyWithValue("cols", cols);

        for (int row = 0; row < array.rows; row++) {
            for (int col = 0; col < array.cols; col++) {
                if ((row == testRow) && (col == testCol)) {
                    assertThat(array.get(row, col)).isOne();
                } else {
                    assertThat(array.get(row, col)).isZero();
                }
            }
        }
    }

    private static Stream<Arguments> provideArrayAndRandomCell() {
        final var randomizer = new Random();
        return IntStream.rangeClosed(1, 10).boxed().flatMap(cols -> {
            return IntStream.rangeClosed(1, 10).boxed().map(rows -> {
                final int testCol = randomizer.nextInt(cols);
                final int testRow = randomizer.nextInt(rows);

                return Arguments.of(cols, rows, testCol, testRow);
            });
        });
    }
}

File: diploma-server/module-common/src/test/kotlin/org/arhor/diploma/commons/IntArray2DTest.kt
package org.arhor.diploma.commons

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import kotlin.random.Random
import kotlin.streams.asStream

internal class IntArray2DTest {

    @ParameterizedTest(name = "should create 2D array of size {0}x{1} filled with {2}")
    @MethodSource("provide rows, cols and initial value")
    fun `IntArray2D#create`(rows: Int, cols: Int, initialValue: Int) {
        // when
        val array = IntArray2D.create(rows, cols, initialValue)

        // then
        assertThat(array).isNotNull
        assertThat(array.rows).isEqualTo(rows)
        assertThat(array.cols).isEqualTo(cols)

        array.forEach { value -> assertThat(value).isEqualTo(initialValue) }
    }

    @ParameterizedTest(name = "should modify only value with [{0}, {1}] index")
    @MethodSource("provide array indices to test")
    fun `IntArray2D#set`(testRow: Int, testCol: Int) {
        // given
        val array = IntArray2D.create(TEST_ROWS, TEST_COLS)

        // when
        array[testRow, testCol] = 1

        // then
        array.forEach { value, row, col ->
            if ((row == testRow) && (col == testCol)) {
                assertThat(value).isOne
            } else {
                assertThat(value).isZero
            }
        }
    }

    @Suppress("UNUSED")
    companion object {
        private const val TEST_ROWS = 5
        private const val TEST_COLS = 5

        @JvmStatic
        fun `provide array indices to test`() = sequence {
            for (testRow in 0 until TEST_ROWS) {
                for (testCol in 0 until TEST_COLS) {
                    yield(Arguments.of(testRow, testCol))
                }
            }
        }.asStream()

        @JvmStatic
        fun `provide rows, cols and initial value`() = sequence {
            for (rows in 1..TEST_ROWS) {
                for (cols in 1..TEST_COLS) {
                    yield(Arguments.of(rows, cols, Random.nextInt(0, 100)))
                }
            }
        }.asStream()
    }
}